### PR TITLE
Add Create Catalog and Edit Catalog functionality

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
This PR adds the capability to Add Catalogs and Edit Catalogs.

Finishes  135552687
Finishes 135552719
Finishes 133964855

image

image

image

image

@miq-bot add_label enhancement
@miq-bot add_label euwe/no

### CODE REVIEW

**Comment by jeff-phillips-18** [comment]
@serenamarie125 @beanh66

File: client/app/components/catalogs/catalog-editor.component.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
+/* eslint camelcase: "off" */
+(function() {
+  'use strict';
+
+  angular.module('app.components')
+    .component('catalogEditor',  {
+      templateUrl: 'app/components/catalogs/catalog-editor.html',
+      bindings: {
+        catalog: "=",
```
Comment (AllenBW) [inline client/app/components/catalogs/catalog-editor.component.js]: No action: must these be two way bound,  or can catalog and serviceTemplates be one way bound (`<`)take advantage of component outputs ('&')?

File: client/app/components/catalogs/catalog-editor.component.js [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+          catalogValid: false,
+          sortTemplates: sortTemplates,
+          handleSave: handleSave,
+          handleCancel: handleCancel,
+          updateSelections: updateSelections,
+          saveCatalog: saveCatalog,
+        }
+      );
+
+      var onDestroy = $scope.$on('$stateChangeStart', onChangeState);
```
Comment (AllenBW) [inline client/app/components/catalogs/catalog-editor.component.js]: No action: can this take advantage of the $onDestroy lifecycle hook?

File: client/app/components/catalogs/catalog-editor.component.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            CatalogsState.addCatalog(response.resources[0].id, removeTemplates).then(removeSuccess, removeFailure);
+          } else {
+            vm.dirty = false;
+            doNavigate();
+          }
+        }
+      }
+
+      function addFailure() {
+        // Remain on the edit page if the save failed
```
Comment (AllenBW) [inline client/app/components/catalogs/catalog-editor.component.js]: No action: should a notification on failed save be throw?

File: client/app/components/catalogs/catalog-list.component.js [hunk 4]
```diff
@@ -9,38 +9,73 @@
...
+        updateCatalogsInfo();
+      });
+
+      function onDestroy() {
+        catalogWatcher();
+        templatesWatcher();
+        tentantsWatcher();
+      }
+
+      $scope.$on('destroy', onDestroy);
```
Comment (AllenBW) [inline client/app/components/catalogs/catalog-list.component.js]: This should use the lifecycle hook `$onDestroy`

File: client/app/components/dual-pane-selector/dual-pane-selector.component.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
+/* eslint camelcase: "off" */
+(function() {
+  'use strict';
+
+  angular.module('app.components')
+    .component('dualPaneSelector', {
+      templateUrl: "app/components/dual-pane-selector/dual-pane-selector.html",
+      bindings: {
+        availableItems: "=",
```
Comment (AllenBW) [inline client/app/components/dual-pane-selector/dual-pane-selector.component.js]: No Action: same comment as above, single bound using outputs

File: client/app/components/dual-pane-selector/dual-pane-selector.html [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  <div class="dual-pane-selector-list-container">
+    <div class="dual-pane-selector-list dual-pane-selector-list-left">
+      <label>{{vm.availableTitle}}</label>
+      <div class="dual-pane-selector-items-container">
+        <div class="dual-pane-selector-item"
+             ng-repeat="item in vm.availableItems track by $index"
+             ng-class="{selected: vm.isAvailableItemSelected(item)}">
+          <div class="dual-pane-selector-info"
+               ng-click="vm.availableItemClick(item)"
+               ng-dblclick="vm.availableItemDoubleClick(item)">
```
Comment (AllenBW) [inline client/app/components/dual-pane-selector/dual-pane-selector.html]: TOO COOL! learning something new every day 🙇‍♀️

File: client/app/services/catalogs-state.service.js [hunk 7]
```diff
@@ -68,6 +68,134 @@
       return CollectionsApi.query('service_templates/' + serviceTemplateId + '/service_dialogs', options);
     };
 
+    catalogState.setCatalogServiceTemplates = function(catalog, serviceTemplates) {
+      if (!catalog.serviceTemplates) {
```
Comment (AllenBW) [inline client/app/services/catalogs-state.service.js]: No action: is this ever false or rather only ever undefined, if so prefer `angular.isUndefined()`

File: client/app/components/catalogs/catalog-editor.component.js (hunk 3 above)
Comment (jeff-phillips-18) [inline client/app/components/catalogs/catalog-editor.component.js]: The service adds the notification.

**Comment by jeff-phillips-18** [comment]
Addressed @AllenBW comments except where noted. Thanks @AllenBW

**Comment by chriskacerguis** [comment]
@serenamarie125 do you need to review, or are you good (b/c it is your team who did this)?

**Comment by beanh66** [comment]
@jeff-phillips-18 Would it be possible to left align the dual pane selector with the other open text boxes for name and description?

**Comment by jeff-phillips-18** [comment]
@beanh66 It is responsive so it changes based on the window width.

**Comment by serenamarie125** [comment]
@chriskacerguis of course we want UX review on everything, regardless of which org the dev is in :) 

@jeff-phillips-18 since the form is embedded on a page, the primary action (Save) should be on the left, and Cancel should be on the right.

**Comment by serenamarie125** [comment]
@jeff-phillips-18 please also tag with "ux" when screens are modified.  Thanks!

**Comment by beanh66** [comment]
@serenamarie125 good catch with save and cancel! @jeff-phillips-18 Sorry I definitely had the buttons switched in my original mocks that you had.

**Comment by jeff-phillips-18** [comment]
@miq-bot add_label ux

**Comment by jeff-phillips-18** [comment]
@serenamarie125  Changed button order for edit screens (Catalogs and Profiles).

**Comment by cshinn** [comment]
LGTM

**Pull request merged**